Commodity/Equity/FX Asian option trade support

gbfredrik opened the pull request:
Hi, Roland and team!

We're happy to share our perhaps biggest implementation, thus far at least, of Asian options in ORE. In short, we link six QL pricing engines to cover discrete/continuous averaging of arithmetic/geometric average price/strike options. Down below I will try to describe the implementation and any design choices made, before I can get to adding it into the user guide as well. There may well be many questionable choices so I will prematurely give a disclaimer that we have many things to learn still :slightly_smiling_face: A few things still remain to be listed, which I will strive to do shortly, but I will share this as a draft to open up for discussion in the meantime. 

We followed suit from the `VanillaOptionTrade` implementation in OREData and created an `AsianOptionTrade` inheriting directly from Trade. From here, we created three further trade type classes: `EquityAsianOption`, `FxAsianOption`, and `CommodityAsianOption`. We use these as a general interface for both average price and average strike options with a single underlying. The specifics of the trade are added to an `OptionAsianData` class inside `OptionData`, with the properties set in the trade XML as:
```xml
<AsianData>
  <!-- Price/Strike -->
  <AsianType>Price</AsianType>
  <!-- Arithmetic/Geometric -->
  <AverageType>Arithmetic</AverageType>
  <!-- All fixing (i.e. averaging) dates -->
  <FixingDates>
    <FixingDate>2021-02-01</FixingDate>
    …
    <FixingDate>2021-02-26</FixingDate>
  </FixingDates>
</AsianData>
```
Note that we opted to list the "fixing dates" individually rather than having a `<StartDate>` and an `<EndDate>`, in order to avoid calendars and such when generating daily/weekly/monthly fixing schedules. I noted this difference in a portfolio set from a project we had previously purchased, though I suppose both alternatives could be implemented in parallell. 

Documentation: Will add descriptions of the trade types to the user guide files, soon. 

Tests: Includes test coverage of the following bits.
- OptionAsianData contructors and inspectors.
- QL::Average::Type string parser
- EquityAsianOption versus literature
- FxAsianOption versus literature
- CommodityAsianOption versus literature
	
Remarks: 
- QL v1.22 added some improved interfaces for Asian options that I would have liked to utilize, however I've delayed doing this currently as I am using an older version of QL (1.18~). 
- We wrap the instrument with the `VanillaInstrument` class even though the class' comments suggests to use some form of an exotic instrument wrapper. As we have only been interested in PVs, we opted to ignore this and sneakily wrapped the Asian instruments with the vanilla wrapper.
- We left the decision of toggling between discrete/continuous averaging to the _pricingengine.xml_ file. Each engine implemented very awkwardly needs to denote if it assumes Continuous or Discrete process type, to determine whether a `DiscreteAveragingAsianOption` or `ContinuousAveragingAsianOption` instrument should be wrapped.
- We've predominantly ran commodity and equity arithmetic average price trades and thus can attest to them working seemingly correct. The FX variant runs fine in tests, at least. As for average (floating) strike options we haven't verified anything other than the parsing, but I assume these could be easily validated with the Asian parity if I'm not mistaken.
- QuantExt holds some items related to APOs (Turnbull-Wakeman/MC engines) but I opted away from these for now - perhaps someone with better insight can share some thoughts on the potential use of those. 


Best regards,
Fredrik Gerdin Börjesson,
SEB

QuaternionRisk commented:
Hi Fredrik, many thanks for your contributions! I have merged all of your non-draft PRs by now. I'd like to get the next release out before mid of July. So I'll review this draft now and get back to you with comments shortly. It would be nice to Asian Options into the release. Best regards, Roland

gbfredrik commented:
Hi Roland! We're happy to see our efforts appreciated. I'll probably have a few additional PRs to send before mid-June, mostly trivial things apart from the FX barrier options. In the meantime as you are reviewing this PR, I'll also try to add some commits with a few minor fixes and refactoring that I think would be valuable. Best regards, Fredrik

QuaternionRisk commented:
Hi Fredrik,
ok, so we went through your design above and in the code and discussed a bit. It is great that you are mobilising the available QuantLib engines and wrote tests that compare to literature, this is very valuable. And the instruments fit pretty well into ORE already, though we believe you could improve the schema somewhat:  
1) Underlying - you are using a "Name" tag  to identify the underlying - could you change this to the "Underlying" that we are using elsewhere in ORE already, see for example EquityOption.
2) AsianData - Peter argued to lift this block one level up in the hierarchy and place it at the same level as OptionData, mainly to keep OptionData slim, the hierarchy more flat and not to complicate schema checks for other products depending on OptionData
3) FixingDates - similarly, we would prefer this block a level up, at the same level as OptionData and AsianData; moreover, can we rename this to ObservationDates? And finally, we would like to keep it flexible, allowing both "Rules" and "Dates" based schedules used elsewhere in the library, i.e.
<ObservationDates>
   <Dates>
      <Date>2021-02-01</Date> 
        ...
      <Date>2021-02-26</Date>
  </Dates>
</ObservationDates>
or
<ObservationDates>
  <Rules>
    <StartDate>...</StartDate>
    <EndDate>...</EndDate>
    <Tenor>...</Tenor>
  </Rules>
</Observationdates>
What do you think? These are fairly superficial changes, not touching on the underlying machinery you have built. I'd be happy to help with these adjustments after merging.
Best regards,
Roland

QuaternionRisk commented:
Hi Fredrik, the XML snippet for the Dates and Rules based schedules following the "i.e."  did not make it into the post above. What I meant is we suggest using a usual ScheduleData structure instead of the explicit list of fixing dates. Best regards, Roland

gbfredrik commented:
Hi Roland, 
I think these are all very good points. I expect I will be able to submit the changes by some time tonight. 

As an addition to the initial post, I also remembered that we successfully used the commodity Asian option type with futures underlyings. Although I don't have any tests for it, the logic is essentially reused from the corresponding vanillas which I think should suffice. 

Another thing to note is that we, in the engine builder, had to purposefully break the observer pattern and extract a single volatility by strike/TTM in order to create `BlackConstantVol` instances on an engine-by-engine basis. This was a workaround to avoid using full volatility surfaces/local volatility in the MC engines as we sometimes saw issues with negative forward variance. I intend to revert this as well, before merging, to use the `CachingOptionEngineBuilder::getBlackScholesProcess(...)` instead. This should help generalize the engine building and avoid breaking the observability. I will run the unit tests after reverting this to make sure nothing unintended occurs. That said, let me know if you'd like to see another approach for the linking of volatility term structures to the engines.

Best regards, Fredrik

gbfredrik commented:
Hi Roland,
I've gotten as far as to implement most of the changes on my side and will finish it up Wednesday. The above three commits are aimed at some of my previous reservations and should have fixed a couple of hidden bugs. I'll mark as ready for review as soon as the other changes are up and get back to you if I have any questions.
Best regards, Fredrik

gbfredrik commented:
Hi Roland, 
I've now implemented the changes suggested after some delays on my side. The tests are updated correspondingly, and you can see the commit message for more detailed notes. 

My only remaining point is regarding the use of the Underlying node that we have in EquityOption as noted. I notice this is not the case for a CommodityOption, nor an FxOption, which is the reason why I opted away from using Underlying for CommodityAsianOption and FxAsianOption. I think it makes sense to use Underlying for those two as well, but then we'd probably also want to update the vanilla equivalents for coherence? That said, I changed the EquityAsianOption trade XML test to use Underlying as I had implemented earlier. 

I'm satisfied with the current status after managing to iron out the previously awkward bits. 
Best regards, Fredrik

QuaternionRisk commented:
Hi Fredrik,
perfect, thanks. And I agree, we should update the vanilla and asian commodity/fx options to accept both the underlying node and "old" name. At least for the vanillas we need to keep the name for backward compatibility. We can add these changes after the merge, and I'll merge this project now.
Best regards,  
Roland

gbfredrik commented:
Hi Roland! Thank you again for helping us with this contribution. I'll see if I can contribute to the documentation of the Asians some time soon, likely after closing a couple of other remaining contributions.
Best regards, Fredrik
